src/md_lammps.c

```diff
@@ -0,0 +1,507 @@
+		return false;
+	}
+
+	data->cell_ext[0] = (float)parse_float(tokens[0]);
+	data->cell_ext[1] = (float)parse_float(tokens[1]);
+	data->cell_ext[2] = (float)parse_float(tokens[2]);
+
+	*/
+
+	return true;
+}
+
+bool md_lammps_data_parse_str(md_lammps_data_t* data, str_t str, struct md_allocator_i* alloc, data_format_t* data_format) {
+	ASSERT(data);
+	ASSERT(alloc);
+
+	md_buffered_reader_t line_reader = md_buffered_reader_from_str(str);
+	return md_lammps_data_parse(data, &line_reader, alloc, data_format);
+}
+
+bool md_lammps_data_parse_file(md_lammps_data_t* data, str_t filename, struct md_allocator_i* alloc, data_format_t* data_format) {
+	bool result = false;
+	md_file_o* file = md_file_open(filename, MD_FILE_READ | MD_FILE_BINARY);
+	if (file) {
+		const int64_t cap = MEGABYTES(1);
+		char* buf = md_alloc(md_heap_allocator, cap);
+
+		md_buffered_reader_t line_reader = md_buffered_reader_from_file(buf, cap, file);
+		result = md_lammps_data_parse(data, &line_reader, alloc, data_format);
+
+		md_free(md_heap_allocator, buf, cap);
+		md_file_close(file);
+	}
+	else {
+		MD_LOG_ERROR("Could not open file '%.*s'", filename.len, filename.ptr);
+	}
+	return result;
+}
+
+void md_lammps_data_free(md_lammps_data_t* data, struct md_allocator_i* alloc) {
+	ASSERT(data);
+	if (data->atom_data) md_array_free(data->atom_data, alloc);
+	MEMSET(data, 0, sizeof(md_lammps_data_t));
+}
+
+bool md_lammps_molecule_init(md_molecule_t* mol, const md_lammps_data_t* data, md_allocator_i* alloc)
+{
+	ASSERT(mol);
+	ASSERT(data);
+	ASSERT(alloc);
+
+	MEMSET(mol, 0, sizeof(md_molecule_t));
+
+	const int64_t num_atoms = data->num_atoms;
+
+	md_array_ensure(mol->atom.x, num_atoms, alloc);
+	md_array_ensure(mol->atom.y, num_atoms, alloc);
+	md_array_ensure(mol->atom.z, num_atoms, alloc);
+	md_array_ensure(mol->atom.residue_idx, num_atoms, alloc);
+
+	int32_t cur_res_id = -1;
+	for (int64_t i = 0; i < num_atoms; ++i) {
+		const float x = data->atom_data[i].x;
+		const float y = data->atom_data[i].y;
+		const float z = data->atom_data[i].z;
+
+		int32_t res_id = data->atom_data[i].mol_idx;
+		if (res_id != cur_res_id) {
+			cur_res_id = res_id;
+			md_residue_id_t id = res_id;
+			md_range_t atom_range = { (uint32_t)mol->atom.count, (uint32_t)mol->atom.count };
+
+			mol->residue.count += 1;
+			md_array_push(mol->residue.id, id, alloc);
+			md_array_push(mol->residue.atom_range, atom_range, alloc);
+		}
+
+		if (mol->residue.atom_range) md_array_last(mol->residue.atom_range)->end += 1;
+
+		mol->atom.count += 1;
+
+		//Set coordinates
+		md_array_push(mol->atom.x, x, alloc);
+		md_array_push(mol->atom.y, y, alloc);
+		md_array_push(mol->atom.z, z, alloc);
+
+		if (mol->residue.count) md_array_push(mol->atom.residue_idx, (md_residue_idx_t)(mol->residue.count - 1), alloc);
+
+		//Set mass
+		const float mass = get_mass(data->atom_type_mass, data->atom_data[i].atom_type, data->num_atom_types);
```

Comment:
If you want, you can use a hash_table to do the lookup. Checkout stb_ds.h

Reply:
Had some issues getting stb_ds hashmap to work. MIght revisit it in the future but for now its not worth putting more time into. Realistically, we can expect just a few mass/type pairs so no need to optimize.